gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/FSJobCatalog.java

```diff
@@ -0,0 +1,288 @@
+      throws Exception {
+    this(jobConfig);
+    this._observerOptional = Optional.of(observer);
+
+    FSPathAlterationListenerAdaptor configFilelistener = new FSPathAlterationListenerAdaptor(this.jobConfDirPath);
+    FSJobCatalogHelper.addPathAlterationObserver(this.pathAlterationDetector, configFilelistener,
+        _observerOptional, this.jobConfDirPath);
+  }
+
+  /**
+   * return the root path of the configuration file folder
+   * The example complete path given a jobSpec is:
+   * [getPath()]/jobSpec.getURI.getRawPath()
+   * @return
+   */
+  public Path getPath() {
+    return this.jobConfDirPath;
+  }
+
+  /**
+   * Fetch all the job files under the jobConfDirPath
+   * @return A collection of JobSpec
+   */
+  @Override
+  public synchronized List<JobSpec> getJobs() {
+    return FSJobCatalogHelper.loadJobConfigHelper(this.jobConfDirPath, Action.BATCHJOB, Optional.<Path>absent());
+  }
+
+  /**
+   * Fetch single job file based on its URI,
+   * return null requested URI not existed
+   * requires null checking
+   * @param uri The relative Path to the target job configuration.
+   * @return
+   */
+  @Override
+  public synchronized JobSpec getJobSpec(URI uri) {
+    Path targetJobSpecPath = new Path(uri);
+
+    List<JobSpec> resultJobSpecList =
+        FSJobCatalogHelper.loadJobConfigHelper(this.jobConfDirPath, Action.SINGLEJOB, Optional.of(targetJobSpecPath));
+    if (resultJobSpecList == null || resultJobSpecList.size() == 0) {
+      LOGGER.warn("No JobSpec with URI:" + uri + " is found.");
+      return null;
+    } else {
+      return resultJobSpecList.get(0);
+    }
+  }
+
+  /**
+   * For each new coming JobCatalogListener, react accordingly to add all existing JobSpec.
+   * @param jobListener
+   */
+  @Override
+  public synchronized void addListener(JobCatalogListener jobListener) {
+    Preconditions.checkNotNull(jobListener);
+
+    this.listeners.addListener(jobListener);
+
+    List<JobSpec> currentJobSpecList = this.getJobs();
+    if (currentJobSpecList == null || currentJobSpecList.size() == 0) {
+      return;
+    } else {
+      for (JobSpec jobSpecEntry : currentJobSpecList) {
+        AddJobCallback addJobCallback = new AddJobCallback(jobSpecEntry);
+        this.listeners.callbackOneListener(addJobCallback, jobListener);
+      }
+    }
+  }
+
+  @Override
+  public synchronized void removeListener(JobCatalogListener jobListener) {
+    this.listeners.removeListener(jobListener);
+  }
+
+  /**
+   * Allow user to programmatically add a new JobSpec.
+   * The method will materialized the jobSpec into real file.
+   *
+   * @param jobSpec The target JobSpec Object to be materialized.
+   *                Noted that the URI return by getUri is a relative path.
+   */
+  @Override
+  public synchronized void put(JobSpec jobSpec) {
+    Preconditions.checkNotNull(jobSpec);
+    try {
+      String tmpStr = jobSpec.getUri().toString();
+      Path tmpPath = new Path(this.jobConfDirPath, tmpStr);
```

Comment:
Suggestion: use `PathUtils.mergePaths(this.jobConfDirPath, new Path(jobSpec.getUri())`